src/main/java/com/google/cloud/pubsublite/kafka/SinglePartitionSubscriber.java

```diff
@@ -0,0 +1,150 @@
+import com.google.cloud.pubsublite.internal.CloseableMonitor;
+import com.google.cloud.pubsublite.internal.wire.Committer;
+import com.google.cloud.pubsublite.proto.SeekRequest;
+import com.google.common.collect.Iterables;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.errorprone.annotations.concurrent.GuardedBy;
+import java.util.ArrayDeque;
+import java.util.Optional;
+
+/** Pulls messages and manages commits for a single partition of a subscription. */
+class SinglePartitionSubscriber {
+  private final PullSubscriberFactory subscriberFactory;
+  private final Partition partition;
+  private final Committer committer;
+
+  private final CloseableMonitor monitor = new CloseableMonitor();
+
+  @GuardedBy("monitor.monitor")
+  private BlockingPullSubscriber subscriber;
+
+  @GuardedBy("monitor.monitor")
+  private boolean needsCommitting = false;
+
+  @GuardedBy("monitor.monitor")
+  private Optional<Offset> lastReceived = Optional.empty();
+
+  SinglePartitionSubscriber(
+      PullSubscriberFactory subscriberFactory,
+      Partition partition,
+      SeekRequest initialSeek,
+      Committer committer)
+      throws CheckedApiException {
+    this.subscriberFactory = subscriberFactory;
+    this.partition = partition;
+    this.committer = committer;
+    this.subscriber =
+        subscriberFactory.newPullSubscriber(partition, initialSeek, this::onSubscriberReset);
+  }
+
+  /** Executes a client-initiated seek. */
+  void clientSeek(SeekRequest request) throws CheckedApiException {
+    try (CloseableMonitor.Hold h = monitor.enter()) {
+      subscriber.close();
+      subscriber = subscriberFactory.newPullSubscriber(partition, request, this::onSubscriberReset);
+    }
+  }
+
+  ApiFuture<Void> onData() {
+    try (CloseableMonitor.Hold h = monitor.enter()) {
+      return subscriber.onData();
+    }
+  }
+
+  @GuardedBy("monitor.monitor")
+  private ArrayDeque<SequencedMessage> pullMessages() throws CheckedApiException {
+    ArrayDeque<SequencedMessage> messages = new ArrayDeque<>();
+    for (Optional<SequencedMessage> message = subscriber.messageIfAvailable();
+        message.isPresent();
+        message = subscriber.messageIfAvailable()) {
+      messages.add(message.get());
+    }
+    return messages;
+  }
+
+  /** Pulls all available messages. */
+  ArrayDeque<SequencedMessage> getMessages() throws CheckedApiException {
+    try (CloseableMonitor.Hold h = monitor.enter()) {
+      ArrayDeque<SequencedMessage> messages = pullMessages();
+      if (!messages.isEmpty()) {
+        lastReceived = Optional.of(Iterables.getLast(messages).offset());
+        needsCommitting = true;
+      }
+      return messages;
+    }
+  }
+
+  Optional<Long> position() {
+    try (CloseableMonitor.Hold h = monitor.enter()) {
+      return lastReceived.map(lastReceived -> lastReceived.value() + 1);
+    }
+  }
+
+  /** Executes a client-initiated commit. */
+  ApiFuture<Void> commitOffset(Offset offset) {
```

Comment:
nit. implement the committer interface as well?

Reply:
There wasn't a need to expose `Committer.waitUntilEmpty`, so I didn't implement the Committer interface.